### PR TITLE
Issue 4235

### LINKED ISSUE #4235: matchShapes and all zeros HuMoments
I've encountered an issue where, when collecting a list of similar shapes using the matchShapes function, I found one shape was being given a perfect match against any shape. Now, after investigation, it turned out that humoments being calculated for the shape matching all were all zero. I'm using a customized version of matchShapes where the hu moments are precalculated, so I can catch these things. For the built in function, this is all done at function level, abstracted away from the user.

I realize this could be reasonably chalked up to a user issue and a lack of sanitizing data before processing it, but I thought I'd bring the issue up here to raise the idea of validating the calculated moments in the function. Otherwise, you'll have a matchShapes function that could, under the right circumstances, match anything to anything.

### PR DESCRIPTION
**Pull request by Tetragramm**
Add check for all zero moments. If one of the shapes is empty, the ma…tch would return zero distance between the shapes even when the other one had content. It now returns DBL_MAX if no moments had value.

### CODE REVIEW

**Comment by vpisarev** [comment]
what about the case when both shapes are empty?

File: modules/imgproc/src/matchcontours.cpp [hunk 1]
```diff
@@ -62,6 +63,11 @@ double cv::matchShapes(InputArray contour1, InputArray contour2, int method, dou
             double ama = fabs( ma[i] );
             double amb = fabs( mb[i] );
 
+            if (ama > 0)
+                anyA = true;
```
Comment (vpisarev) [inline modules/imgproc/src/matchcontours.cpp]: it would be more elegant to write `anyA = ama > 0;`

Comment (Tetragramm) [inline modules/imgproc/src/matchcontours.cpp]: No, that's not the same. That can set anyA from true to false, which we don't want.

`anyA |= ama >0` would work, but then you have to depend on people remember what |= does, and noticing the |, which a quick skim would probably miss.

**Comment by vpisarev** [comment]
@Tetragramm, you are right! Looking at the patch (just the changed lines) I did not realize that this code `if(...) any... = true;` is inside the loop. :+1:

**Pull request merged**